### PR TITLE
prowlarr: init at 0.1.1.978

### PR DESCRIPTION
**Pull request by jdreaver**
###### Motivation for this change

https://github.com/NixOS/nixpkgs/issues/135887

This PR adds a new package and NixOS service for [Prowlarr](https://wiki.servarr.com/prowlarr). This is an indexer similar to `jackett` (which is already in nixpkgs), but is newer and more well-integrated with the *arr ecosystem (e.g. `radarr` and `sonarr`, which are also already in `nixpkgs`).

I mostly copied the config from `radarr` and fixed the `prowlarr`-specific bits. In particular, the `update.sh` script had to be modified because the releases are not official releases yet, so we have to specify the version we want to use.

**This is my first significant `nixpkgs` PR.** I've updated the versions of packages before, but I've never added a new service from scratch. I would think anyone who is a maintainer for `radarr` could/should review this. I'm also happy to add myself as a maintainer for this package, but I'm not sure what the protocol on that is.

I used this `nixpkgs` fork and enabled `prowlarr` in my home server's NixOS config, and it worked flawlessly, so this has basically been tested end-to-end.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jdreaver** [comment]
I meant to open this against my fork, sorry for the noise.

**Comment by jdreaver** [comment]
Okay, I tested this end-to-end and it is ready for review!

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 136039` at ab21e02c run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>prowlarr</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/servers/prowlarr/default.nix:47:3:
```
   |
47 |   meta = with lib; {
   |   ^
```
</li>
  </ul>
</details>

File: nixos/modules/services/misc/prowlarr.nix [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    networking.firewall = mkIf cfg.openFirewall {
+      allowedTCPPorts = [ 9696 ];
+    };
+
+    users.users = mkIf (cfg.user == "prowlarr") {
+      prowlarr = {
+        group = cfg.group;
+        home = cfg.dataDir;
+        uid = config.ids.uids.prowlarr;
```
Comment (aanderse) [inline nixos/modules/services/misc/prowlarr.nix:67]: Please use `isSystemUser` instead. https://github.com/NixOS/rfcs/blob/master/rfcs/0052-dynamic-ids.md for details.

File: nixos/modules/services/misc/prowlarr.nix [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      };
+    };
+
+    users.groups = mkIf (cfg.group == "prowlarr") {
+      prowlarr.gid = config.ids.gids.prowlarr;
```
Comment (aanderse) [inline nixos/modules/services/misc/prowlarr.nix:72]: Please drop this line.

**Review by aanderse — CHANGES_REQUESTED** [review]

**Comment by jdreaver** [comment]
Thanks for the review @aanderse! I fixed those issues, and I also added myself as a maintainer for the package.

**Review by aanderse — APPROVED** [review]
Module LGTM :+1:

**Comment by aanderse** [comment]
@jdreaver I don't generally review packages, only modules. Maybe @SuperSandro2000 would spare a few minutes to review the package so we can merge.

**Comment by jdreaver** [comment]
I rebased against master to fix a merge conflict in the release notes.

@SuperSandro2000 do you have time to review this, or do you know anyone else who would have time?

**Comment by jdreaver** [comment]
Hey @aanderse and @SuperSandro2000, do you know how I could get the module reviewed? Is there some way to flag this PR as needing review? Does @aanderse's approval mark it as not needing review somehow?

**Comment by aanderse** [comment]
@jdreaver please try this out: https://github.com/timokau/marvin-mk2#as-a-pr-author

If this hasn't been merged in 7 days please ping again.

**Comment by jdreaver** [comment]
Thanks for the tip @aanderse, and thanks for showing me that Discourse thread!

/marvin opt-in
/status needs_reviewer

File: nixos/modules/services/misc/prowlarr.nix [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        Type = "simple";
+        User = cfg.user;
+        Group = cfg.group;
+        ExecStart = "${pkgs.prowlarr}/bin/Prowlarr -nobrowser -data='${cfg.dataDir}'";
+        Restart = "on-failure";
+      };
+    };
+
+    networking.firewall = mkIf cfg.openFirewall {
+      allowedTCPPorts = [ 9696 ];
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/prowlarr.nix:60]: Is the port not configurable?

File: pkgs/servers/prowlarr/default.nix [hunk 4]
```diff
@@ -0,0 +1,54 @@
+{ lib, stdenv, fetchurl, mono, libmediainfo, sqlite, curl, makeWrapper, icu, dotnetCorePackages, openssl, nixosTests }:
+
+let
+  os = if stdenv.isDarwin then "osx" else "linux";
```
Comment (SuperSandro2000) [inline pkgs/servers/prowlarr/default.nix:4]: We can't assume that else it is linux. It should throw if it is not darwin or linux.

File: nixos/modules/services/misc/prowlarr.nix (hunk 3 above)
Comment (jdreaver) [reply nixos/modules/services/misc/prowlarr.nix:60]: It looks to me that all of the *arr apps (sonarr, radarr, prowlarr, etc) are currently configured _only_ with their UIs. I have a personal TODO item to figure out how to configure them with nix. They are C# apps that either use XML or SQLite (or something else, I can't remember off the top of my head), so we'll have to figure out how to inject our configuration into a format usable by those apps.

TL;DR: no, this isn't currently configurable easily outside of the prowlarr UI.

File: pkgs/servers/prowlarr/default.nix (hunk 4 above)
Comment (jdreaver) [reply pkgs/servers/prowlarr/default.nix:4]: Ah agreed, sounds good!

**Comment by jdreaver** [comment]
Thanks for the review @SuperSandro2000! This is ready for another look when you have time.

File: nixos/tests/prowlarr.nix [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import ./make-test-python.nix ({ lib, ... }:
+
+with lib;
+
+{
+  name = "prowlarr";
+  meta.maintainers = with maintainers; [ ];
```
Comment (ambroisie) [inline nixos/tests/prowlarr.nix:7]: Add yourself as a maintainer.

Comment (jdreaver) [reply nixos/tests/prowlarr.nix:7]: ah good catch, thanks!

File: nixos/modules/services/misc/prowlarr.nix (hunk 3 above)
Comment (mweinelt) [reply nixos/modules/services/misc/prowlarr.nix:60]: Then we probably shouldn't offer this option, if the port can be brought out of sync from within the webui.

Comment (ambroisie) [reply nixos/modules/services/misc/prowlarr.nix:60]: @mweinelt I disagree, for three reasons. One is consistency: the other *arr services all have a version of this option made available to the user if they want to. The second one is discoverability: as a user, if I enable the service, I have no way to know which port I should connect to when first setting-up my proxy/the software. Finally, if someone enables the option, they would know that changing the port in the webui would obviously break their setup.

Comment (jdreaver) [reply nixos/modules/services/misc/prowlarr.nix:60]: Agreed with @ambroisie on both points (discoverability and consistency).

Comment (mweinelt) [reply nixos/modules/services/misc/prowlarr.nix:60]: Not a strong opinion here, I'm a network guy by trade and coarse firewalling makes me itchy. I'm not going to use the `openFirewall` rule anyway.

**Comment by mweinelt** [comment]
Requested a review from maintainers of sonarr, bazarr, lidarr and radarr. Maybe one of them feels comfortable merging this.

File: nixos/modules/services/misc/prowlarr.nix (hunk 3 above)
Comment (jdreaver) [reply nixos/modules/services/misc/prowlarr.nix:60]: Hah, yeah I with NixOS firewalls were different. I am toying with a single nftables module in my configs instead of the builtin firewall. I understand how the builtin firewall works, but I want my rules to be really clear when I query them, and not have a ton of empty NixOS-specific chains.

File: pkgs/servers/prowlarr/update.sh [hunk 6]
```diff
@@ -0,0 +1,55 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i bash -p curl gnused nix-prefetch jq
+
+set -e
```
Comment (SuperSandro2000) [inline pkgs/servers/prowlarr/update.sh:4]:
```suggestion
set -eou pipefail
```

File: pkgs/servers/prowlarr/default.nix [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  nativeBuildInputs = [ makeWrapper ];
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/{bin,share/${pname}-${version}}
+    cp -r * $out/share/${pname}-${version}/.
+
+    makeWrapper "${dotnetCorePackages.netcore_3_1}/bin/dotnet" $out/bin/Prowlarr \
+      --add-flags "$out/share/${pname}-${version}/Prowlarr.dll" \
```
[the remark this answers is not in the document]
Comment (mweinelt) [reply pkgs/servers/prowlarr/default.nix:42]: Why?

**Review by purcell — APPROVED** [review]
This honestly looks fine to me, either with or without the latest suggested changes from @SuperSandro2000.

File: nixos/modules/services/misc/prowlarr.nix [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+    systemd.services.prowlarr = {
+      description = "Prowlarr";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        Type = "simple";
+        User = cfg.user;
+        Group = cfg.group;
```
Comment (mweinelt) [inline nixos/modules/services/misc/prowlarr.nix:53]: Can we remove `dataDir`, `user` and `group` in favor of `DynamicUser`? We probably don't need a static uid/gid setup and `StateDirectory` takes care of permissions, so you could remove the tmpfiles rule as well.

```
DynamicUser = true;
User = "prowlarr";
StateDirectory = "prowlarr";
```

**Comment by mweinelt** [comment]
Awesome, so let's get this PR in order. To merge I need:

- 1 commit `maintainers: add <yourname>`
- 1 commit `prowlarr: init at <version>` for the package
- 1 commit `nixos/prowlarr: init` for the module
  - the release note changes belongs to the module
- 1 commit `nixos/tests/prowlarr: init` for the test
  - in which you add the `passthru.tests` option to the package

And let's discuss if we can simplify this alot with `DynamicUser` and `StateDirectory`. This is systemd hardening 101. :wink:

**Comment by mweinelt** [comment]
The commit structure looks good now!

**Review by mweinelt — APPROVED** [review]
Thanks, tested and works as expected!

**Comment by jdreaver** [comment]
@mweinelt ah good catch, this is probably a symptom of me copy/pasting too liberally and then modifying the script :sweat_smile: 

I ran the script as-is in that directory, like `./update.sh 0.1.1.978`. It sounds like this won't work as a `passthru.updateScript` until the issue in this comment is resolved

https://github.com/jdreaver/nixpkgs/blob/3b98f4851f256759851f864a5e39328f495640f9/pkgs/servers/prowlarr/update.sh#L36-L49

```
# N.B. Prowlarr is still in development, so
# https://api.github.com/repos/Prowlarr/Prowlarr/releases/latest
# returns nothing. Instead, we require the release as an argument. Go
# to https://github.com/Prowlarr/Prowlarr/releases to see releases.
```

Do you happen to know a workaround? Should I remove this script as a `passthru.updateScript` for now?

**Comment by mweinelt** [comment]
```
git ls-remote --tags git@github.com:Prowlarr/Prowlarr.git | cut -d '/' -f3  | sort --version-sort | tail -n1
```

**Comment by jdreaver** [comment]
@mweinelt I found a similar solution that uses the API and `jq` to sort the versions as strings. Here is the diff:

```diff
modified   pkgs/servers/prowlarr/update.sh
@@ -3,14 +3,6 @@
 
 set -eou pipefail
 
-if [ "$#" -ne 1 ]; then
-    echo "Usage: $0 VERSION" >&2
-    echo "Example: $0 0.1.1.875" >&2
-    exit 1
-fi
-
-version=$1
-
 dirname="$(dirname "$0")"
 
 updateHash()
@@ -33,23 +25,25 @@ updateVersion()
     sed -i "s/version = \"[0-9.]*\";/version = \"$1\";/g" "$dirname/default.nix"
 }
 
+currentVersion=$(cd $dirname && nix eval --raw -f ../../.. prowlarr.version)
+
 # N.B. Prowlarr is still in development, so
 # https://api.github.com/repos/Prowlarr/Prowlarr/releases/latest
-# returns nothing. Instead, we require the release as an argument. Go
-# to https://github.com/Prowlarr/Prowlarr/releases to see releases.
-
-# currentVersion=$(cd $dirname && nix eval --raw -f ../../.. prowlarr.version)
+# returns nothing. Once this endpoint returns something, we should use
+# it. Until then, we use jq to sort releases as strings, which is
+# error-prone, so please verify the latest version used!
 
 # latestTag=$(curl https://api.github.com/repos/Prowlarr/Prowlarr/releases/latest | jq -r ".tag_name")
 # latestVersion="$(expr $latestTag : 'v\(.*\)')"
+latestVersion=$(curl https://api.github.com/repos/Prowlarr/Prowlarr/git/refs/tags | jq '. | map(.ref) | sort | .[-1]' -r | sed 's|refs/tags/v||')
 
-# if [[ "$currentVersion" == "$latestVersion" ]]; then
-#     echo "Prowlarr is up-to-date: ${currentVersion}"
-#     exit 0
-# fi
+if [[ "$currentVersion" == "$latestVersion" ]]; then
+    echo "Prowlarr is up-to-date: ${currentVersion}"
+    exit 0
+fi
 
-updateVersion $version
+updateVersion $latestVersion
 
-updateHash $version x64 linux
-updateHash $version arm64 linux
-updateHash $version x64 osx
+updateHash $latestVersion x64 linux
+updateHash $latestVersion arm64 linux
+updateHash $latestVersion x64 osx
```

I manually changed the file version and hashes to dummy files, and I got the script to work as a passthru:

Before:

```diff
diff --git a/pkgs/servers/prowlarr/default.nix b/pkgs/servers/prowlarr/default.nix
index 511217a48f1..caa8fbe2900 100644
--- a/pkgs/servers/prowlarr/default.nix
+++ b/pkgs/servers/prowlarr/default.nix
@@ -16,14 +16,14 @@ let
   }."${stdenv.hostPlatform.system}" or (throw "Unsupported system: ${stdenv.hostPlatform.system}");
 
   hash = {
-    x64-linux_hash = "sha256-9DoqyotXAUha2TMSgDIot5PD8ABpfZ8gsshS1ypr5SY=";
-    arm64-linux_hash = "sha256-r22c70OuevRsF8gOHZOkkhlRtoD4nsTHnXF82elQIF8=";
-    x64-osx_hash = "sha256-6jVM4iSGT7tpagocI/1nuBPVvAegfFqsCfrz2fPKCI4=";
+    x64-linux_hash = "sha256-FAKE";
+    arm64-linux_hash = "sha256-FAKE";
+    x64-osx_hash = "sha256-FAKE";
   }."${arch}-${os}_hash";
 
 in stdenv.mkDerivation rec {
   pname = "prowlarr";
-  version = "0.1.1.978";
+  version = "0.0.1";
 
   src = fetchurl {
     url = "https://github.com/Prowlarr/Prowlarr/releases/download/v${version}/Prowlarr.develop.${version}.${os}-core-${arch}.tar.gz";
```

```
$ nix-shell maintainers/scripts/update.nix --argstr package prowlarr
this derivation will be built:
  /nix/store/z2f9yfxpk9bhr0vg0gm5ls4ca4i9mgwa-packages.json.drv
building '/nix/store/z2f9yfxpk9bhr0vg0gm5ls4ca4i9mgwa-packages.json.drv'...

Going to be running update for following packages:
 - prowlarr-0.0.1

Press Enter key to continue...

Running update for:
 - prowlarr-0.0.1: UPDATING ...
 - prowlarr-0.0.1: DONE.

Packages updated!
```

```
$ git diff
$
```

**Comment by jdreaver** [comment]
Okay one last change so we sort by version numbers properly instead of sorting them as raw strings:

```diff
diff --git a/pkgs/servers/prowlarr/update.sh b/pkgs/servers/prowlarr/update.sh
index 9fb40da46c7..d61edda5a0a 100755
--- a/pkgs/servers/prowlarr/update.sh
+++ b/pkgs/servers/prowlarr/update.sh
@@ -30,12 +30,13 @@ currentVersion=$(cd $dirname && nix eval --raw -f ../../.. prowlarr.version)
 # N.B. Prowlarr is still in development, so
 # https://api.github.com/repos/Prowlarr/Prowlarr/releases/latest
 # returns nothing. Once this endpoint returns something, we should use
-# it. Until then, we use jq to sort releases as strings, which is
-# error-prone, so please verify the latest version used!
+# it. Until then, we use jq to sort releases (N.B. the "sort_by(. |
+# split(".") | map(tonumber))" incantation is to sort the version
+# number properly and not as a string).
 
 # latestTag=$(curl https://api.github.com/repos/Prowlarr/Prowlarr/releases/latest | jq -r ".tag_name")
 # latestVersion="$(expr $latestTag : 'v\(.*\)')"
-latestVersion=$(curl https://api.github.com/repos/Prowlarr/Prowlarr/git/refs/tags | jq '. | map(.ref) | sort | .[-1]' -r | sed 's|refs/tags/v||')
+latestVersion=$(curl https://api.github.com/repos/Prowlarr/Prowlarr/git/refs/tags | jq '. | map(.ref | sub("refs/tags/v";"")) | sort_by(. | split(".") | map(tonumber)) | .[-1]' -r)
 
 if [[ "$currentVersion" == "$latestVersion" ]]; then
     echo "Prowlarr is up-to-date: ${currentVersion}"
```

**Comment by mweinelt** [comment]
And that's it. Thanks!
